### PR TITLE
oc: update route warnings for oc status

### PR DESCRIPTION
**Pull request by kargakis**
@deads2k @ramr @liggitt 

Fixes https://github.com/openshift/origin/issues/6266

### CODE REVIEW

**Comment by kargakis** [comment]
[test]

**Comment by deads2k** [comment]
@stevekuznetsov ptal

File: pkg/route/generator/generate.go [hunk 1]
```diff
@@ -77,8 +78,14 @@ func (RouteGenerator) Generate(genericParams map[string]interface{}) (runtime.Ob
...
 	if len(portString) > 0 {
+		var targetPort intstr.IntOrString
+		if port, err := strconv.Atoi(portString); err == nil {
+			targetPort = intstr.FromInt(port)
+		} else {
+			targetPort = intstr.FromString(portString)
+		}
 		route.Spec.Port = &api.RoutePort{
-			TargetPort: intstr.FromString(portString),
+			TargetPort: targetPort,
```
Comment (ramr) [inline pkg/route/generator/generate.go]: Any reason to do the Atoi here?  `intstr.IntValue()` does that automatically if type is String, which is what it is when  TargetPort is set to `intstr.FromString(portString)`.

Comment (kargakis) [inline pkg/route/generator/generate.go]: Yes, `IntValue` returns a number independent of what `portString` is. I would have to check if that number is zero in order to determine if `portString` is a number or a string. Instead, by doing the conversion directly I know on the spot what I need to use (`FromInt` vs `FromString`).

Comment (ramr) [inline pkg/route/generator/generate.go]: So what I see is a check here just to invoke `intstr.From{Int,String}' and in the code further down (https://github.com/kargakis/origin/blob/route-status-fixes/pkg/route/graph/analysis/analysis.go#L80), there is still a call to the `intstr.String()` method and the `Atoi` conversion again [which is fine - we should check if its a string or an int at that point]. Seems like this conversion here is not really needed - just using `intstr.FromString()` would work here.

Comment (kargakis) [inline pkg/route/generator/generate.go]: It would work but since route port is a IntOrString, I would prefer to save the correct type in it (inststr.Int in case of an int, intstr.String in case of a string) and not go with [string type](https://github.com/kubernetes/kubernetes/blob/f046d6c83e82d48e5f32539bae677661f3c505c4/pkg/util/intstr/intstr.go#L61) all the time.

The code in the oc status marker is deliberately like that to handle cases where we have saved int types as strings (what this code was doing before) but it is wrong, right?

File: pkg/cmd/cli/describe/projectstatus.go [hunk 2]
```diff
@@ -338,7 +338,7 @@ func getMarkerScanners(logsCommandName, securityPolicyCommandFormat, setProbeCom
 		func(g osgraph.Graph, f osgraph.Namer) []osgraph.Marker {
 			return deployanalysis.FindDeploymentConfigReadinessWarnings(g, f, setProbeCommandName)
 		},
-		routeanalysis.FindMissingPortMapping,
+		routeanalysis.FindPortMappingIssues,
```
Comment (stevekuznetsov) [inline pkg/cmd/cli/describe/projectstatus.go]: Might be nice to have more smaller, more targeted methods, but it's a small nit.

File: pkg/route/graph/analysis/analysis.go [hunk 3]
```diff
@@ -52,6 +56,7 @@ route:
 					Key:      MissingServiceWarning,
 					Message: fmt.Sprintf("%s is supposed to route traffic to %s but %s doesn't exist.",
 						f.ResourceName(routeNode), f.ResourceName(svcNode), f.ResourceName(svcNode)),
+					// TODO: Suggest 'oc create service' once that's a thing.
```
Comment (stevekuznetsov) [inline pkg/route/graph/analysis/analysis.go]: Is there an issue open for this?

File: pkg/route/graph/analysis/analysis.go [hunk 4]
```diff
@@ -70,6 +75,54 @@ route:
 
 				continue route
 			}
+
+			if routeNode.Spec.Port != nil {
+				routePortString := routeNode.Spec.Port.TargetPort.String()
+				if routePort, err := strconv.Atoi(routePortString); err == nil {
+					// route has numeric port, service has no port with that number as a targetPort.
```
Comment (stevekuznetsov) [inline pkg/route/graph/analysis/analysis.go]: second part of this comment is misleading and only valid after the following loop

File: pkg/route/graph/analysis/analysis.go [hunk 5]
```diff
@@ -70,6 +75,54 @@ route:
...
+					}
+
+					if len(svcNode.Spec.Ports) == 1 {
+						marker.Suggestion = osgraph.Suggestion(fmt.Sprintf("oc patch %s -p '{\"spec\":{\"port\":{\"targetPort\": %d}}}'", f.ResourceName(routeNode), svcNode.Spec.Ports[0].TargetPort.IntValue()))
+					}
+					markers = append(markers, marker)
+
+					continue route
+				} else {
+					// route has named targetPort, service has no port with that name.
```
Comment (stevekuznetsov) [inline pkg/route/graph/analysis/analysis.go]: ditto about comment here

File: pkg/route/graph/analysis/analysis.go [hunk 6]
```diff
@@ -70,6 +75,54 @@ route:
...
 				continue route
 			}
+
+			if routeNode.Spec.Port != nil {
+				routePortString := routeNode.Spec.Port.TargetPort.String()
+				if routePort, err := strconv.Atoi(routePortString); err == nil {
+					// route has numeric port, service has no port with that number as a targetPort.
+					for _, port := range svcNode.Spec.Ports {
+						if port.TargetPort.IntValue() == routePort {
+							continue route
```
Comment (stevekuznetsov) [inline pkg/route/graph/analysis/analysis.go]: named `continue` targets smell of lots of cyclomatic complexity, maybe we'd be better off with breaking this function up and getting rid of the need for them

File: pkg/cmd/cli/describe/projectstatus.go (hunk 2 above)
Comment (kargakis) [inline pkg/cmd/cli/describe/projectstatus.go]: All of the names used in marking are similar. If we feel that there is room for improvement we should do it in a separate PR for all of them.

File: pkg/route/graph/analysis/analysis.go (hunk 3 above)
Comment (kargakis) [inline pkg/route/graph/analysis/analysis.go]: There is https://github.com/kubernetes/kubernetes/pull/19509 upstream which hopefully will turn into what we want.

File: pkg/route/graph/analysis/analysis.go (hunk 4 above)
Comment (kargakis) [inline pkg/route/graph/analysis/analysis.go]: Right, moving after the loop...

File: pkg/route/graph/analysis/analysis.go (hunk 5 above)
Comment (kargakis) [inline pkg/route/graph/analysis/analysis.go]: Right, moving after the loop...

File: pkg/route/graph/analysis/analysis.go (hunk 6 above)
Comment (kargakis) [inline pkg/route/graph/analysis/analysis.go]: Comment in the same fashion as https://github.com/openshift/origin/pull/7833#discussion_r55297690. We could break these down to separate self-contained units but continuing has served us well so far. If we decide that it's a real problem, we should refactor all marking functions.

File: pkg/route/generator/generate.go (hunk 1 above)
Comment (kargakis) [inline pkg/route/generator/generate.go]: > Actually since everything is clientside I am going to refactor the oc status part of the code to not use the conversion but decide based on the type. WDYT?

ugh, actually I just remembered why I couldn't do this in the first place. There may be existing routes (these coming from `oc expose` :) ) that use a route port with a string type but an int value...

File: pkg/route/graph/analysis/analysis.go [hunk 7]
```diff
@@ -70,6 +76,56 @@ route:
 
 				continue route
 			}
+
+			if routeNode.Spec.Port == nil {
+				continue route
```
Comment (deads2k) [inline pkg/route/graph/analysis/analysis.go]: We've got too many of these now.  Refactor the inner loop into a method and use `return`.

File: pkg/route/graph/analysis/analysis.go [hunk 8]
```diff
@@ -70,6 +76,56 @@ route:
...
+			}
+
+			for _, port := range svcNode.Spec.Ports {
+				if port.Name == routePortString {
+					continue route
+				}
+			}
+
+			// route has named targetPort, service has no port with that name.
+			markers = append(markers, osgraph.Marker{
```
Comment (deads2k) [inline pkg/route/graph/analysis/analysis.go]: Can't you do the same logic for a suggestion that you used above?  If you have a service with exactly one port, they could use a patch.  I liked that.

**Comment by deads2k** [comment]
Minor comments.  I like it.

File: pkg/route/generator/generate.go (hunk 1 above)
Comment (ramr) [inline pkg/route/generator/generate.go]: ok, thx that looks good.

**Comment by kargakis** [comment]
comments addressed [merge]
